tree/ntuple/v7/src/RNTupleSerialize.cxx

```diff
@@ -247,13 +247,21 @@ std::uint32_t SerializePhysicalColumn(const ROOT::Experimental::RColumnDescripto
 std::uint32_t SerializeColumnsOfFields(const ROOT::Experimental::RNTupleDescriptor &desc,
                                        std::span<const ROOT::Experimental::DescriptorId_t> fieldList,
                                        const ROOT::Experimental::Internal::RNTupleSerializer::RContext &context,
-                                       void *buffer)
+                                       void *buffer, bool forHeaderExtension)
 {
    auto base = reinterpret_cast<unsigned char *>(buffer);
    auto pos = base;
    void **where = (buffer == nullptr) ? &buffer : reinterpret_cast<void **>(&pos);
 
+   const auto *xHeader = !forHeaderExtension ? desc.GetHeaderExtension() : nullptr;
+
    for (auto parentId : fieldList) {
+      // If we're serializing the non-extended header and we already have a header extension (which may happen if
+      // we load an RNTuple for incremental merging), we need to skip all the extended fields, as they need to be
+      // written in the header extension, not in the regular header.
+      if (xHeader && xHeader->ContainsField(parentId))
```

Comment:
Don't we need to check if the _column_ is part of the extension header? Then we would also cover extended column representations.

Reply:
I need to check both, right? Since the header extension only contains information about the extended column ids..

Reply:
I thought that you'd anyway only process the first $k$ fields of the regular header if `forHeaderExtension` is false.